File: sdk/python/examples/durable_run.py
"""Application core for one durable Colossus run.

Connection and credential loading stay in trusted application composition. Pass an
already connected ``ColossusClient``; never move its bearer into argv, an environment
variable, a descriptor, or a renderer.
"""

from __future__ import annotations

import uuid
from collections.abc import Awaitable, Callable, Iterable
from dataclasses import dataclass
from typing import Protocol, runtime_checkable

from colossus.api.v1alpha1 import agent_run_pb2, session_pb2

from colossus_sdk import ColossusClient, decode_colossus_rpc_error

InteractionHandler = Callable[
    [agent_run_pb2.Interaction],
    Awaitable[agent_run_pb2.RespondInteractionRequest | None],
]


@runtime_checkable
class RpcErrorLike(Protocol):
    """Minimal error surface required by the bounded rich-status decoder."""

    def trailing_metadata(self) -> Iterable[tuple[str, str | bytes]] | None: ...

    def code(self) -> object: ...


@dataclass(frozen=True)
class DurableRunResult:
    """Released terminal result plus bounded tool activity seen along the way."""

    run_id: str
    output: str
    tool_names: tuple[str, ...]


class DurableRunFailed(RuntimeError):
    """A terminal run failure with retry and outcome metadata preserved."""

    def __init__(self, failure: agent_run_pb2.RunFailure) -> None:
        super().__init__(failure.message)
        self.reason = failure.reason
        self.recoverable = failure.recoverable
        self.outcome_certainty = failure.outcome_certainty
        self.http_status = failure.http_status if failure.HasField("http_status") else None
        self.retry_after_ms = failure.retry_after_ms if failure.HasField("retry_after_ms") else None


async def run_prompt(
    client: ColossusClient,
    prompt: str,
    *,
    mode: agent_run_pb2.RunMode = agent_run_pb2.RUN_MODE_EXECUTE,
    max_turns: int = 12,
    handle_interaction: InteractionHandler | None = None,
) -> DurableRunResult:
    """Create once, watch durably, and return only released terminal output."""

    request = agent_run_pb2.CreateRunRequest(
        input=[session_pb2.ContentPart(text=session_pb2.TextContent(text=prompt))],
        role="primary",
        mode=mode,
        max_turns=max_turns,
        idempotency_key=f"sdk-example-create-{uuid.uuid4()}",
    )
    try:
        created = await client.agent_runs.create_run(request)
    except Exception as error:
        detail = decode_colossus_rpc_error(error) if isinstance(error, RpcErrorLike) else None
        if detail is not None:
            raise RuntimeError(
                f"CreateRun failed: {detail.reason}; retryable={detail.retryable}; "
                f"outcome={detail.outcome_certainty}"
            ) from error
        raise

    run_id = created.run.run_id
    tool_names: set[str] = set()
    async for response in client.agent_runs.watch_run(run_id):
        update = response.update
        update_case = update.WhichOneof("update")
        if update_case == "tool_activity":
            tool_names.add(update.tool_activity.tool_name)
        elif update_case == "interaction":
            if not update.interaction.respondable_by_caller:
                continue
            if handle_interaction is None:
                raise RuntimeError(
                    "run is waiting for an interaction; provide handle_interaction and "
                    "resume from the last durable cursor"
                )
            interaction_response = await handle_interaction(update.interaction)
            if interaction_response is not None:
                await client.agent_runs.respond_interaction(interaction_response)
        elif update_case == "result":
            return DurableRunResult(run_id, update.result.output, tuple(sorted(tool_names)))
        elif update_case == "failure":
            raise DurableRunFailed(update.failure.failure)
        elif update_case == "cancellation":
            raise RuntimeError(f"run cancelled: {update.cancellation.message}")

    raise RuntimeError("run watch ended without an exact terminal update")


def deny_approval(
    interaction: agent_run_pb2.Interaction,
) -> Awaitable[agent_run_pb2.RespondInteractionRequest | None]:
    """Example safe default: deny approval requests and leave prompts unanswered."""

    async def decide() -> agent_run_pb2.RespondInteractionRequest | None:
        if interaction.WhichOneof("content") != "approval":
            return None
        return agent_run_pb2.RespondInteractionRequest(
            run_id=interaction.run_id,
            interaction_id=interaction.interaction_id,
            etag=interaction.etag,
            idempotency_key=f"sdk-example-interaction-{uuid.uuid4()}",
            approval_answer=agent_run_pb2.ApprovalAnswer(
                approved=False,
                request_hash=interaction.approval.request_hash,
            ),
        )

    return decide()

File: sdk/typescript/examples/durable-run.ts
/**
 * Application core for one durable Colossus run.
 *
 * Compose this with createSecureGrpcClient() in trusted native/server code. The
 * bearer credential must come from application-protected storage and must never be
 * placed in argv, environment variables, descriptors, logs, or renderer memory.
 */

import { randomUUID } from "node:crypto";

import type { ServiceError } from "@grpc/grpc-js";

import {
  decodeColossusRpcError,
  isTerminalRunUpdate,
  watchRun,
} from "../src/index.js";
import {
  type AgentRunServiceClient,
  type CreateRunResponse,
  type GetRunResponse,
  type Interaction,
  type RespondInteractionRequest,
  type RespondInteractionResponse,
  RunMode,
  type RunUpdate,
  type WatchRunResponse,
} from "../src/gen/colossus/api/v1alpha1/agent_run.js";

export interface DurableRunResult {
  readonly runId: string;
  readonly output: string;
  readonly toolNames: readonly string[];
}

export type InteractionHandler = (
  interaction: Interaction,
) => Promise<RespondInteractionRequest | undefined>;

export class DurableRunFailed extends Error {
  public readonly reason: string;
  public readonly recoverable: boolean;
  public readonly outcomeCertainty: number;
  public readonly httpStatus: number | undefined;
  public readonly retryAfterMs: bigint | undefined;

  public constructor(failure: NonNullable<RunUpdate["update"]> & {
    readonly $case: "failure";
  }) {
    const detail = failure.value.failure;
    super(detail?.message ?? "run failed without released failure detail");
    this.name = "DurableRunFailed";
    this.reason = detail?.reason ?? "run.failure_detail_missing";
    this.recoverable = detail?.recoverable ?? false;
    this.outcomeCertainty = detail?.outcomeCertainty ?? 0;
    this.httpStatus = detail?.httpStatus;
    this.retryAfterMs = detail?.retryAfterMs;
  }
}

export async function runPrompt(
  client: AgentRunServiceClient,
  prompt: string,
  options: {
    readonly mode?: RunMode;
    readonly maxTurns?: number;
    readonly signal?: AbortSignal;
    readonly handleInteraction?: InteractionHandler;
  } = {},
): Promise<DurableRunResult> {
  let created: CreateRunResponse;
  try {
    created = await unary<CreateRunResponse>((callback) =>
      client.createRun(
        {
          input: [{ content: { $case: "text", value: { text: prompt } } }],
          role: "primary",
          mode: options.mode ?? RunMode.RUN_MODE_EXECUTE,
          selectedSkills: [],
          maxTurns: options.maxTurns ?? 12,
          idempotencyKey: `sdk-example-create-${randomUUID()}`,
        },
        callback,
      ),
    );
  } catch (error) {
    const detail =
      error !== null &&
      typeof error === "object" &&
      "code" in error &&
      "metadata" in error
        ? decodeColossusRpcError(
            error as Pick<ServiceError, "code" | "metadata">,
          )
        : undefined;
    if (detail !== undefined) {
      throw new Error(
        `CreateRun failed: ${detail.reason}; retryable=${String(detail.retryable)}; ` +
          `outcome=${detail.outcomeCertainty}`,
        { cause: error },
      );
    }
    throw error;
  }
  const runId = created.run?.runId;
  if (runId === undefined || runId.length === 0) {
    throw new Error("CreateRun returned no durable run identity");
  }

  const toolNames = new Set<string>();
  const feed = watchRun<RunUpdate>({
    runId,
    ...(options.signal === undefined ? {} : { signal: options.signal }),
    open: async function* (watchedRunId, afterSequence, signal) {
      const stream = client.watchRun({
        runId: watchedRunId,
        afterSequence,
      });
      const cancel = (): void => stream.cancel();
      signal?.addEventListener("abort", cancel, { once: true });
      try {
        for await (const response of stream as AsyncIterable<WatchRunResponse>) {
          if (response.update === undefined) {
            throw new Error("WatchRun returned an empty response");
          }
          yield {
            runId: response.update.runId,
            sequence: response.update.sequence,
            value: response.update,
          };
        }
      } finally {
        signal?.removeEventListener("abort", cancel);
      }
    },
    reconcile: async (watchedRunId) => {
      const response = await unary<GetRunResponse>((callback) =>
        client.getRun({ runId: watchedRunId }, callback),
      );
      if (response.run === undefined) {
        throw new Error("GetRun returned no run during watch reconciliation");
      }
      return {
        runId: response.run.runId,
        lastSequence: response.run.lastSequence,
        terminal: response.run.terminal !== undefined,
      };
    },
    isTerminal: isTerminalRunUpdate,
  });

  for await (const item of feed) {
    const update = item.value.update;
    switch (update?.$case) {
      case "toolActivity":
        toolNames.add(update.value.toolName);
        break;
      case "interaction": {
        if (!update.value.respondableByCaller) {
          break;
        }
        const request = await options.handleInteraction?.(update.value);
        if (request === undefined) {
          throw new Error(
            "run is waiting for an interaction; supply handleInteraction and resume " +
              "from the last durable cursor",
          );
        }
        await unary<RespondInteractionResponse>((callback) =>
          client.respondInteraction(request, callback),
        );
        break;
      }
      case "result":
        return {
          runId,
          output: update.value.output,
          toolNames: [...toolNames].sort(),
        };
      case "failure":
        throw new DurableRunFailed(update);
      case "cancellation":
        throw new Error(`run cancelled: ${update.value.message}`);
      default:
        break;
    }
  }
  throw new Error("run watch ended without an exact terminal update");
}

export async function denyApproval(
  interaction: Interaction,
): Promise<RespondInteractionRequest | undefined> {
  if (interaction.content?.$case !== "approval") {
    return undefined;
  }
  return {
    runId: interaction.runId,
    interactionId: interaction.interactionId,
    etag: interaction.etag,
    idempotencyKey: `sdk-example-interaction-${randomUUID()}`,
    response: {
      $case: "approvalAnswer",
      value: {
        approved: false,
        requestHash: interaction.content.value.requestHash,
      },
    },
  };
}

function unary<Response>(
  invoke: (callback: (error: ServiceError | null, response: Response) => void) => unknown,
): Promise<Response> {
  return new Promise((resolve, reject) => {
    invoke((error, response) => {
      if (error !== null) {
        reject(error);
      } else {
        resolve(response);
      }
    });
  });
}
